### ISSUE TITLE
UTF-8 letters support

### ISSUE THREAD

**Issue by lavrenkov-sketch** [ultralytics/yolov5#4351]
How to add support for UTF-8 fonts, I want to display predictions in Russian, the opencv library itself seems to be able to do this

**Comment by kinoute** [comment]
I'm interested too. The `plot_one_box` function uses OpenCV to draw rectangles and put texts. By default OpenCV seems limited when it comes to fonts (no unicode version) but they added in version 3 a `free type` module to load custom true type font:

See: http://fireant.github.io/misc/2017/01/28/ttf-opencv.html

There is `plot_one_box_PIL` in `utils/plots.py` that uses PIL instead. PIL can load custom True Type fonts but I couldn't make it work with `detect.py` and  "Arial Unicode" or another custom font. Rectangles are not displayed either.

**Comment by kinoute** [comment]
By the way, not only labels containing UTF-8 characters such as accents don't appear on images, the results (`class_name(s)`), seem to remove the accents as well. The problem doesn't appear only when calling openCV, it seems general to the project.

Example: "Société France" would become "Socit France"

**Comment by glenn-jocher** [comment]
@kinoute @lavrenkov-sketch we have a PIL-based plotter that might be better for UTF-8:
https://github.com/ultralytics/yolov5/blob/e96c74b5a1c4a27934c5d8ad52cde778af248ed8/utils/plots.py#L82-L94

**Comment by lavrenkov-sketch** [comment]
@glenn-jocher how can i use plot_one_box_PIL it returns an array and doesn't draw any labels on the image

**Comment by lavrenkov-sketch** [comment]
im0 = plot_one_box_PIL(xyxy, im0, label=label, color=colors(c, True), line_thickness=line_thickness)

**Comment by glenn-jocher** [comment]
@lavrenkov-sketch I've opened up PR #4407 to assign a default line width with PIL plotting, not sure why it was missing. 

Can you supply an example class name that is failing with the existing cv2 plotting function? We should update this to switch to PIL automatically on detection of UTF-8 characters.

**Comment by kinoute** [comment]
"Société Générale", image from W&B (same in inference, running the model through Docker Hub):

![Capture d’écran 2021-08-13 à 17 21 34](https://user-images.githubusercontent.com/623763/129381770-ada6b294-435f-4786-b892-93fd7ad019d7.png)

**Comment by glenn-jocher** [comment]
@kinoute perfect thanks! I'll see if we can merge a more comprehensive fix this weekend.

But wait in this case the cv2 label function works with a silent error of not displaying the é characters? Is this master code or did you modify this to remove these characters?

**Comment by kinoute** [comment]
I didn't modify anything, it comes from a training I did 3 days ago. I
train with the official docker image.

To avoid this problem, I temporarily removed the accents from my yaml file.
Do note that accents don't show up in "results" at inference time too. It's
not only a CV2 problem.

**Comment by glenn-jocher** [comment]
@kinoute I see correct console printout but incorrect cv2 plotting in Colab.

<img width="744" alt="Screenshot 2021-08-13 at 23 14 13" src="https://user-images.githubusercontent.com/26833433/129419102-0a97e4dc-695e-4f7c-8433-96f2030b870a.png">

Can you run this code to reproduce in you environment:
```python
# PyTorch Hub
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')
model.names[0] = 'Société Générale'

# Images
dir = 'https://ultralytics.com/images/'
imgs = [dir + f for f in ('zidane.jpg', 'bus.jpg')]  # batch of images

# Inference
results = model(imgs)
results.print()
results.save()
results.pandas().xyxy[0]
```

**Comment by kinoute** [comment]
@glenn-jocher It works like this because you overwrite the class name right before the inference, but I think the problem comes from the time the classes are loaded from the yaml dataset file. I was about to reproduce the issue like this:

```shell
# download yolov5 master
git clone https://github.com/ultralytics/yolov5
cd yolov5

# download coco128 dataset
chmod +x data/scripts/get_coco128.sh && ./data/scripts/get_coco128.sh

# change first class name in data/coco128.yaml from "person" to "person Société Générale"
vi data/coco128.yaml
```

![Capture d’écran 2021-08-14 à 00 13 20](https://user-images.githubusercontent.com/623763/129423104-f0600821-32c7-44af-ac4d-2150e513c2a3.png)

```shell
# train
python train.py --img 640 --batch 16 --epochs 3 --data coco128.yaml --weights yolov5s.pt —cache
```

After the training, we can confirm some images have accents removed, such as `runs/train/exp/confusion_matrix.png` and `runs/train/exp/val_batch0_labels.jpg`:

![Capture d’écran 2021-08-14 à 00 15 10](https://user-images.githubusercontent.com/623763/129423298-ca74aff7-6ad9-474c-a43c-70c67d6e68ea.png)
![Capture d’écran 2021-08-14 à 00 15 28](https://user-images.githubusercontent.com/623763/129423302-69d96c73-98d2-4faf-bac3-8f5be6e64088.png)

Now, on the detection/inference, running:

```shell
python detect.py --weights runs/train/exp/weights/best.pt --img 640 --conf 0.25 --source data/images/ --save-txt
```

shows the same problem too:
![Capture d’écran 2021-08-14 à 00 18 12](https://user-images.githubusercontent.com/623763/129423407-fa5be707-ebdc-49d9-a0ee-5183684a376b.png)

And finally, using the Docker Hub for inference:

```shell
 /sites/yolov5  python3
Python 3.8.9 (default, Apr  3 2021, 01:49:17)
[Clang 11.0.0 (clang-1100.0.33.17)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> model = torch.hub.load('ultralytics/yolov5', 'custom', path="runs/train/exp/weights/best.pt")
Using cache found in /Users/kinoute/.cache/torch/hub/ultralytics_yolov5_master
YOLOv5 🚀 2021-8-14 torch 1.7.0 CPU

Fusing layers...
Model Summary: 224 layers, 7266973 parameters, 0 gradients, 17.0 GFLOPs
Adding AutoShape...
>>> dir = 'https://ultralytics.com/images/'
>>> imgs = [dir + f for f in ('zidane.jpg', 'bus.jpg')]  # batch of images
>>> results = model(imgs)
>>> results.print()
image 1/2: 720x1280 2 person Socit Gnrales, 2 ties
image 2/2: 1080x810 4 person Socit Gnrales, 1 bus
Speed: 536.4ms pre-process, 403.4ms inference, 2.1ms NMS per image at shape (2, 3, 640, 640)
>>> results.save()
Saved 2 images to 'runs/detect/exp3'
>>> results.pandas().xyxy[0]
         xmin        ymin         xmax        ymax  confidence  class  \
0  748.832520   38.970703  1150.665039  711.768921    0.866193      0
1  113.337891  193.695984  1127.441406  712.344666    0.676592      0
2  433.699280  433.693848   517.761475  715.654663    0.659638     27
3  986.345520  303.328003  1027.942627  420.201050    0.278434     27

                  name
0  person Socit Gnrale
1  person Socit Gnrale
2                  tie
3                  tie
>>>
```

We can see in the "textual" result, the accents don't appear as well.

**Comment by glenn-jocher** [comment]
@kinoute good news 😃! Your original issue may now be partially fixed ✅ in PR #4413. This PR removes 'ascii' encoding that had been in place when reading yaml files, so now the UTF-8 characters are saved correctly in the model names list.

Unfortunately the cv2 plotting issue remains though, which will require separate additional fixes. In this example I trained a new model with an updated coco128.yaml as in your above example. The model best.pt correctly saves the UTF-8 characters:

<img width="1553" alt="Screenshot 2021-08-14 at 13 38 27" src="https://user-images.githubusercontent.com/26833433/129445291-3e6fd43f-a402-43c7-9866-12aba3f5740a.png">

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by glenn-jocher** [comment]
TODO: UTF-8 image plotting fix

**Comment by kinoute** [comment]
I will have a look this afternoon and report to you it the fix works, thanks!

Regarding PIL UTF-8, I think the Nvidia docker image is based on Ubuntu 20.04. We could switch from Arial.ttf to another built-in font such as the Ubuntu font, which is unicode by default. But we have to find a way to deal with people that don't run Yolov5 through the docker image as well.

I think there is a way to fix CV2 too, but it will need way more work and changes.

**Comment by glenn-jocher** [comment]
@kinoute I think we need to merge the PIL and cv2 plotting functions into one, which accepts a `use_pil` boolean, which is set to True automatically if UTF characters are found. See https://stackoverflow.com/questions/54189911/cv2-imwrite-and-german-letters-%C3%A4-%C3%BC-%C3%B6

**Comment by glenn-jocher** [comment]
@lavrenkov-sketch @kinoute good news 😃! Your original issue may now be completely resolved ✅ in PR #4416. This PR fixes UTF image annotations, and combines with PR #4413, which fixes UTF console output, to fully resolve UTF usage with YOLOv5.

<img width="1400" alt="Screenshot 2021-08-14 at 17 17 09" src="https://user-images.githubusercontent.com/26833433/129450991-5bceb7e2-7363-497f-875e-9c500de63782.png">

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
